# surf-2017
This repository contains the work for 2017 Summer Undergraduate Research Fellowship Project in the Kwiatkowski Lab at Brigham and Women's Hospital. In this project, I developed a novel analysis pipeline for identifying mutations in DNA sequencing data at a user-specified allele frequency and optimized this program using a multithreading-based solution.

### Abstract: ###
>Massively parallel sequencing (MPS) is a form of sequencing technology for high-throughput DNA sequencing projects, including the analysis of human genetic variants. In the Kwiatkowski Lab, MPS is used to detect sequence variants and mutations in patients who have the genetic disorder Tuberous Sclerosis Complex, and in several types of tumors and cancers. Using MPS, the output received in the lab is a binary alignment map (BAM) file containing reference genome-aligned sequencing data, as well as a collection of reads which do not align correctly. However, BAM files alone are not feasible for identifying variation found in the input DNA. Rather, they must be interpreted and analyzed by other programs to obtain data that indicates DNA sequence variants. We have developed a robust, user-friendly program written in the Unix command line, Python, and MATLAB that specifically searches the BAM file data for sequence variants across a given gene at a user-specified allele frequency and determines potential sites of large-scale deletions. This program has been tested with various projects and will be used on the labâ€™s entire collection of BAM files from various genomic and cell-free DNA samples.
